Allow configuring per-mount-point per-queue-type disk alarms

the-mikedavis opened the pull request:
This is an extension of the free disk space alarm which allows configuring additional mount points to monitor and which queue type(s) to block when they are near full. For example with a config like so:

```ini
stream.data_dir = /mnt/data/streams
# Directory where the file system is mounted.
disk_free_limits.stream.mount_point = /mnt/data/streams
# Alarm threshold: if free space falls under this absolute
# limit then an alarm fires per queue type.
disk_free_limits.stream.absolute = 2GB
# Queue types to block when the threshold is breached.
disk_free_limits.stream.queue_types = stream
```

Publishers to streams would be blocked once the free space of `/data/stream-data` falls under 2GB. Publishers to classic or quorum queues could continue though.

The motivation of this feature is that you may want to use separate disks for different queue types. For example for higher throughput you may want to use volume(s) with better throughput and/or IOPS for streaming but use standard disks for queue data. Also, alarms are currently fairly aggressive by blocking all publishing. Ideally you should be able to continue using queues when the space you have allocated for streams fills up, or vice versa.

This is a different approach than https://github.com/rabbitmq/rabbitmq-server/pull/14086. Instead of measuring disk usage under a directory like [`du(1)`](https://www.man7.org/linux/man-pages/man1/du.1.html), `rabbit_disk_monitor` is updated to measure free space of all mounts at once with [`disksup:get_disk_info/0`](https://www.erlang.org/doc/apps/os_mon/disksup#get_disk_info/0). Under the hood this performs the same [`df(1)`](https://www.man7.org/linux/man-pages/man1/df.1.html) check as `rabbit_disk_monitor` had been doing previously - measuring mount-point free space is much cheaper than measuring directory disk footprint. Monitoring mount points is also quite flexible: you can use multiple disks on one mount point with RAID-0 striping or split up a single disk with partitions.

This is a draft - it needs tests and currently only AMQP 0-9-1 is updated to perform selecting blocking. All other protocols currently block for any alarm.

Some of the commits in this branch are refactors that could be cherry-picked out. https://github.com/rabbitmq/rabbitmq-server/pull/14814 is pretty trivial and the refactors to use maps instead of `dict` in `rabbit_alarm` and use `disksup` instead of the custom `df` code in `rabbit_disk_alarm` are not strictly related to the feature here.

Discussed in https://github.com/rabbitmq/rabbitmq-server/discussions/14590

the-mikedavis committed:
rabbit_env: Enable disksup in os_mon but set threshold to 1.0

Previously we set `start_disksup` to `false` to avoid OTP's automatic
monitoring of disk space. `disksup`'s gen_server starts a port (which
runs `df` on Unix) which measures disk usage and sets an alarm through
OTP's `alarm_handler` when usage exceeds the configured
`disk_almost_full_threshold`. We can set this threshold to 1.0 to
effectively turn off disksup's monitoring (i.e. the alarm will never be
set).

By enabling disksup we have access to `get_disk_data/0` and
`get_disk_info/0,1` which can be used to replace the copied versions in
`rabbit_disk_monitor`.

the-mikedavis committed:
rabbit_disk_monitor: Use disksup to determine available bytes

`disksup` now exposes the calculation for available disk space for a
given path using the same `df` mechanism on Unix. We can use this
directly and drop the custom code which reimplements that.

the-mikedavis committed:
rabbit.schema: Add config options for per-queue-type disk limits

the-mikedavis committed:
rabbit_disk_monitor: Monitor per-queue-type mounts

the-mikedavis committed:
rabbit_alarm: Prefer maps to dicts

This is not a functional change, just a refactor to eliminate dicts and
use maps instead. This cleans up some helper functions like
dict_append/3, and we can use map comprehensions in some places to
avoid intermediary lists.

the-mikedavis committed:
rabbit_alarm: Add a helper to format resource alarm sources

Co-authored-by: Sunny Katkuri <skatkur@amazon.com>

the-mikedavis committed:
Set per-queue-type disk alarms for configured mounts

This introduces a new variant of `rabbit_alarm:resource_alarm_source()`:
`{disk, QueueType}` which triggers when the configured mount for queue
type(s) fall under their limit of available space.

the-mikedavis committed:
AMQP 0-9-1: Handle per-queue-type disk alarms

This covers both network and direct connections for 0-9-1. We store a
set of the queue types which have been published into on both a channel
and connection level since blocking is done on the connection level but
only the channel knows what queue types have been published.

Then when the published queue types or the set of alarms changes, the
connection evaluates whether it is affected by the alarm. If not it may
publish but once a channel publishes to an alarmed queue type the
connection then blocks until the channel exits or the alarm clears.

the-mikedavis commented on deps/amqp_client/src/amqp_gen_connection.erl:
```diff
@@ -214,18 +215,30 @@ handle_cast({register_blocked_handler, HandlerPid},
+        false ->
+            {noreply, State1}
+    end;
+handle_cast({channel_published_to_queue_type, _ChPid, QT},
```
This feature might need a feature flag. Here for direct connections if old client code is used on a newer server then it would error after publishing since it isn't expecting this cast. I think it would be unlikely to happen in practice but the mixed-version test suite will probably run into this.

the-mikedavis committed:
prometheus: Add core metrics per-mount for free space and limit

This adds two gauge metrics which are emitted per configured mount, one
for available bytes and the other for the low watermark. The label
`"disk=<name>"` is attached to both gauges to distinguish which mount
the gauge applies to.

Co-authored-by: Luke Bakken <lrbakken@amazon.com>

the-mikedavis committed:
management: Expose current mount stats in API and UI

This adds the configured mounts, if there are any, to the API JSON
response for the `/api/nodes` and `/api/node/<node>` endpoints and to
the overview and node-detail UI pages. Time series data is not collected
for these metrics - that should be scraped from the Prometheus endpoint
instead.

the-mikedavis committed:
rabbit_disk_monitor: Add config to tune polling interval

The polling interval (min, max and fast-rate) should be tuned
for use on different hardware. For example high-end machines with strong
network bandwidth should be tuning the fast-rate higher so that disk
space is checked more often, as with stronger resources the disk space
could fill up more rapidly than the default 250MB/sec predicts.

Co-authored-by: Luke Bakken <lrbakken@amazon.com>

the-mikedavis committed:
CLI: Extend set_disk_free_limit to set mount limits

With this change you can say:

    rabbitmqctl set_disk_limit mount Streaming 2GiB

This applies the limit only to the "Streaming" mount.

the-mikedavis committed:
Allow configuring classic queue data dir in Cuttlefish config

This is the same kind of toggle as the Raft data directory and stream
data directory but for controlling classic queue data location.

the-mikedavis committed:
rabbit_stream_reader: Block during stream queue-type disk alarm

the-mikedavis committed:
MQTT: Handle per-queue-type disk alarms

This includes regular MQTT and MQTT-over-WebSockets.

the-mikedavis committed:
AMQP 1.0: Handle per-queue-type disk alarms

lukebakken committed:
rabbit_disk_monitor: Document NaN alarm behavior

When disk information is unavailable, the 'available' field in mount
records is set to 'NaN'. Due to Erlang term ordering, 'NaN' < Limit
evaluates to true, which triggers alarms for those mounts.

This is correct fail-safe behavior - when we cannot determine available
disk space, we block publishing to prevent potential disk exhaustion.

Add comments to alarmed_mounts/1 and alarmed_queue_types/1 explaining
this intentional behavior.

lukebakken committed:
unit_disk_monitor_SUITE: Test that duplicate mount paths are allowed

Two disk limit entries may share the same mount point with different
limits and queue types. For example, a higher limit for streams and a
lower limit for classic queues on the same disk is a valid and useful
configuration.

Add a schema test that verifies this configuration is accepted.

lukebakken commented on deps/rabbitmq_prometheus/src/collectors/prometheus_rabbitmq_core_metrics_collector.erl:
```diff
@@ -865,6 +869,12 @@ get_data(exchange_names, _, _) ->
                         [{Label, 1}|Acc]
                 end, [], rabbit_exchange:list());
+get_data(mount_metrics, _, _) ->
+    [{<<"disk=", Name/binary>>, Available, Limit}
```
```suggestion
    [{<<"disk=\"", Name/binary, "\"">>, Available, Limit}
```

lukebakken reviewed (commented):
All other Prometheus label values are in double-quotes, so it seems like `disk` should be as well.

lukebakken committed:
rabbit.schema: Validate that disk_free_limits entries have all required fields

A partial disk_free_limits entry (e.g. only name and mount, missing limit
and queue_types) would pass schema validation but cause a function_clause
error in rabbit_disk_monitor:mounts/0 at node startup.

Add validation in the disk_free_limits translation that checks all four
required fields are present for each entry, and add a cross-reference
comment in rabbit_disk_monitor:mounts/0.

Add a test that verifies an incomplete entry is rejected with an
appropriate error message.

lukebakken committed:
rabbit_alarm: Clarify log message for per-queue-type disk alarms

The existing 'Publishers will be blocked until this alarm clears' message
is preserved unchanged for all alarm types to avoid breaking log monitoring
that depends on this text.

For {disk, QueueType} alarms, add a second log line clarifying that only
publishers to that specific queue type are blocked, not all publishers.

Extract the shared log message into log_resource_alarm_set/2 to avoid
duplication.

lukebakken committed:
rabbit_disk_monitor: Handle multiple disks in resolve_data_dir/0

When disksup:get_disk_info/1 returns multiple mount points for the data
directory path, pick the most specific one (longest path) rather than
returning an error. All returned mount points are prefix matches of the
data directory, so the longest path is always the correct choice.

Refactor resolve_data_dir/0 to delegate to a testable resolve_data_dir/1
that accepts the disk info list directly, and export it for testing.

Add unit tests for the single result and multiple results cases.

lukebakken reviewed (commented):
Here's a summary of the review performed by @lukebakken and the 🧞:

## Commits added during review

- f534680fb8 - `rabbit_disk_monitor`: Document NaN alarm behavior in `alarmed_mounts/1` and `alarmed_queue_types/1`
- 82da3af0fb - `unit_disk_monitor_SUITE`: Test that duplicate mount paths are allowed
- b5587f6b17 - `rabbit.schema`: Validate that `disk_free_limits` entries have all required fields; test added
- ddb65ff3bb - `prometheus`: Fix mount metric label to use quoted value (`disk="name"` not `disk=name`)
- f1c68778e7 - `set_disk_free_limit_command_test`: Add tests for the `mount` subcommand
- c051d42f58 - `rabbit_alarm`: Clarify log message for per-queue-type disk alarms
- f1379d179f - `rabbit_prometheus_http_SUITE`: Test mount metric label format
- a23840ac92 - `rabbit_disk_monitor`: Handle multiple disks in `resolve_data_dir/0`
- b8a3c72117 - `amqp_client_SUITE`: Test per-queue-type disk alarm blocking for direct connections
- 66f298e441 - `amqp_client`: Per-channel `queue_types_published` tracking for direct connections

## Issues resolved

**`amqp_gen_connection.erl`**: The original flat set approach meant a direct connection that once published to a queue type would always be considered as publishing to that type, even after channels closed. This has been fixed: the server channel pid is now threaded through the channel startup path as a tagged tuple `{direct, ServerPid} | {network, WriterPid}`, allowing `amqp_gen_connection` to track `queue_types_published` per server channel pid and clean up correctly when a channel closes. The full `amqp_client_SUITE` (157 tests) passes with this change.

**Duplicate queue types**: If the same queue type appears in two `disk_free_limits` entries, `alarmed_queue_types/1` deduplicates via `sets:union` - no crash, no double-alarming. A connection is blocked if any mount carrying that queue type is alarmed, and unblocked only when all such mounts clear. This is correct and intentional behavior.

## Issues to discuss

**STOMP and STOMP-over-WebSockets**: These protocols are architecturally different from all others: they are built on top of `amqp_client` rather than implementing the broker protocol directly. `rabbit_stomp_processor` uses `amqp_connection:start(#amqp_params_direct{})`, so publishes go through `amqp_gen_connection` → `rabbit_channel`. This creates a layering problem:

- `amqp_gen_connection` is registered with `rabbit_alarm` (via `rabbit_direct`) and now correctly implements selective per-queue-type blocking
- `rabbit_stomp_reader` is **also** independently registered with `rabbit_alarm` and has its own socket throttling via `control_throttle`, which uses `not sets:is_empty(BlockedBy)` - blocking on any alarm regardless of queue type
- `rabbit_web_stomp_handler` has a `conserve_resources` boolean field that appears to be dead code - nothing in the codebase sends `{conserve_resources, Conserve}` to it

The correct fix is to remove `rabbit_stomp_reader`'s direct `rabbit_alarm` registration and instead register it as the block handler on `amqp_gen_connection` after login, letting `amqp_gen_connection` drive socket throttling via `connection.blocked` / `connection.unblocked`. This is pre-existing behavior not introduced by this PR - STOMP was already blocking on all alarms before. Is this worth addressing in this PR or should it be a follow-up?

## Notes

- The PR description says "currently only AMQP 0-9-1 is updated to perform selective blocking" - this appears to be outdated. Stream, MQTT, MQTT-over-WebSockets, and AMQP 1.0 all have selective blocking implemented in the branch.

lukebakken committed:
amqp_client_SUITE: Test per-queue-type disk alarm blocking for direct connections

Add a test that verifies a direct connection (as used by plugins such as
STOMP) is blocked when a per-queue-type disk alarm fires and the connection
has published to that queue type, and unblocked when the alarm clears.

Note: direct connections use a flat set for queue_types_published with no
per-channel cleanup. Once a connection publishes to a queue type it is
permanently considered as publishing to that type for the lifetime of the
connection. This is conservative but correct: the client and server channel
pids live on different nodes with no existing correlation mechanism.

lukebakken committed:
amqp_client: Per-channel queue_types_published tracking for direct connections

Previously, queue_types_published in amqp_gen_connection was a flat set
with no per-channel cleanup. Once a connection published to a queue type
it was permanently considered as publishing to that type, causing
over-blocking when channels closed.

Fix by threading the server channel pid through the channel startup path
as a tagged tuple {direct, ServerPid} | {network, WriterPid}:

- amqp_channel_sup: start_writer/6 returns a tagged writer tuple;
  start_link/6 passes it through to the caller
- amqp_channel: set_writer/2 unwraps the tagged tuple
- amqp_channels_manager: stores ClientPid -> TaggedWriter mapping;
  sends channel_closed/2 with the tagged writer on channel exit
- amqp_gen_connection: tracks queue_types_published per server channel
  pid (keyed by {direct, ServerPid}); removes the entry on channel_closed

Network connections are unaffected: rabbit_reader handles blocking
independently and amqp_gen_connection ignores {network, _} channel_closed.

Test: amqp_client_SUITE:per_queue_type_disk_alarm_direct_connection
verifies that a direct connection is unblocked when the publishing
channel closes, even while the alarm remains active.
